tests/ranged_balance.cpp

```diff
@@ -0,0 +1,317 @@
+    }
+    dispersion.add_range( shooter.recoil );
+
+    return dispersion;
+}
+
+static void test_shooting_scenario( npc &shooter, int min_quickdraw_range,
+                                    int min_good_range, int max_good_range )
+{
+    {
+        dispersion_sources dispersion = get_dispersion( shooter, 0 );
+        std::array<statistics, 5> minimum_stats = firing_test( dispersion, min_quickdraw_range, {{ 0.2, 0.1, -1, -1, -1 }} );
+        INFO( dispersion );
+        INFO( "Range: " << min_quickdraw_range );
+        INFO( "Max aim speed: " << shooter.aim_per_move( shooter.weapon, MAX_RECOIL ) );
+        INFO( "Min aim speed: " << shooter.aim_per_move( shooter.weapon, shooter.recoil ) );
+        CAPTURE( minimum_stats[0].n() );
+        CAPTURE( minimum_stats[0].adj_wald_error() );
+        CAPTURE( minimum_stats[1].n() );
+        CAPTURE( minimum_stats[1].adj_wald_error() );
+        CHECK( minimum_stats[0].avg() < 0.2 );
+        CHECK( minimum_stats[1].avg() < 0.1 );
+    }
+    {
+        dispersion_sources dispersion = get_dispersion( shooter, 225 );
+        std::array<statistics, 5> good_stats = firing_test( dispersion, min_good_range, {{ -1, -1, 0.5, -1, -1 }} );
+        INFO( dispersion );
+        INFO( "Range: " << min_good_range );
+        INFO( "Max aim speed: " << shooter.aim_per_move( shooter.weapon, MAX_RECOIL ) );
+        INFO( "Min aim speed: " << shooter.aim_per_move( shooter.weapon, shooter.recoil ) );
+        CAPTURE( good_stats[2].n() );
+        CAPTURE( good_stats[2].adj_wald_error() );
+        CHECK( good_stats[2].avg() > 0.5 );
+    }
+    {
+        dispersion_sources dispersion = get_dispersion( shooter, 300 );
+        std::array<statistics, 5> good_stats = firing_test( dispersion, max_good_range, {{ -1, -1, 0.1, -1, -1 }} );
+        INFO( dispersion );
+        INFO( "Range: " << max_good_range );
+        INFO( "Max aim speed: " << shooter.aim_per_move( shooter.weapon, MAX_RECOIL ) );
+        INFO( "Min aim speed: " << shooter.aim_per_move( shooter.weapon, shooter.recoil ) );
+        CAPTURE( good_stats[2].n() );
+        CAPTURE( good_stats[2].adj_wald_error() );
+        CHECK( good_stats[2].avg() < 0.1 );
+    }
+}
+
+static void test_fast_shooting( npc &shooter, int moves, float hit_rate )
+{
+    const int fast_shooting_range = 3;
+    dispersion_sources dispersion = get_dispersion( shooter, moves );
+    std::array<statistics, 5> fast_stats = firing_test( dispersion, fast_shooting_range, {{ -1, hit_rate, -1, -1, -1 }} );
+    INFO( dispersion );
+    INFO( "Range: " << fast_shooting_range );
+    INFO( "Max aim speed: " << shooter.aim_per_move( shooter.weapon, MAX_RECOIL ) );
+    INFO( "Min aim speed: " << shooter.aim_per_move( shooter.weapon, shooter.recoil ) );
+    CAPTURE( fast_stats[1].n() );
+    CAPTURE( fast_stats[1].adj_wald_error() );
+    CHECK( fast_stats[1].avg() > hit_rate );
+}
+
+void assert_encumbrance( npc &shooter, int encumbrance )
```

Comment:
Not an issue, just suggestion for the future:
Since encumbrance tests are often used, it would be useful to have a way of ensuring exact encumbrance just by number. For example, calling item factory to generate a set of temporary items with exact desired stats, providing player class with some `encumbrance_cache` structure which could be injected with a derived class, or creating a temporary trait/set of traits that does the job.

Tying encumbrance to real items in a test can be a problem when someone wants to change stats of those items and suddenly tests fail. Especially since they're really random here: beekeeping gloves?